File: src/main/java/io/klbz/curie/Either.java
package io.klbz.curie;

import java.util.Objects;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

import static io.klbz.curie.Maybe.just;
import static io.klbz.curie.Maybe.none;

public interface Either<L, R> {
	static <L, R> Either<L, R> left(L value){ return new Left<>(value); }

	static <L, R> Either<L, R> right(R value){ return new Right<>(value); }

	Either<R, L> swap();

	<LPrime> Either<LPrime, R> mapL(Function<? super L, LPrime> f);

	<RPrime> Either<L, RPrime> mapR(Function<? super R, RPrime> f);

	<LPrime> Either<LPrime, R> flatMapL(Function<? super L, Either<LPrime, R>> f);

	<RPrime> Either<L, RPrime> flatMapR(Function<? super R, Either<L, RPrime>> f);

	L collapseIntoL(Function<? super R, ? extends L> f);

	R collapseIntoR(Function<? super L, ? extends R> f);

	Maybe<L> isolateL();

	Maybe<R> isolateR();

	boolean satisfiesL(Predicate<? super L> p);

	boolean satisfiesR(Predicate<? super R> p);

	boolean satisfies(Predicate<? super L> lp, Predicate<? super R> rp);

	Either<L, R> whenLDo(Consumer<L> doF);

	Either<L, R> whenRDo(Consumer<R> doF);

	L assumeL();

	L assumeL(Supplier<? extends RuntimeException> toThrowWhenAssumptionInvalid);

	R assumeR();

	R assumeR(Supplier<? extends RuntimeException> toThrowWhenAssumptionInvalid);

	enum Alternative {Left, Right}

	class Left<L, R> implements Either<L, R> {
		private final L value;

		private Left(L value){ this.value = value; }

		@Override
		public Either<R, L> swap(){ return right(value); }

		@Override
		public <LPrime> Either<LPrime, R> mapL(Function<? super L, LPrime> f){ return left(f.apply(value)); }

		@Override
		public <RPrime> Either<L, RPrime> mapR(Function<? super R, RPrime> f){ return left(value); }

		@Override
		public <LPrime> Either<LPrime, R> flatMapL(Function<? super L, Either<LPrime, R>> f){ return f.apply(value); }

		@Override
		public <RPrime> Either<L, RPrime> flatMapR(Function<? super R, Either<L, RPrime>> f){ return left(value); }

		@Override
		public L collapseIntoL(Function<? super R, ? extends L> f){ return value; }

		@Override
		public R collapseIntoR(Function<? super L, ? extends R> f){ return f.apply(value); }

		@Override
		public Maybe<L> isolateL(){ return just(value); }

		@Override
		public Maybe<R> isolateR(){ return none(); }

		@Override
		public boolean satisfiesL(Predicate<? super L> p){ return p.test(value); }

		@Override
		public boolean satisfiesR(Predicate<? super R> p){ return false; }

		@Override
		public boolean satisfies(Predicate<? super L> lp, Predicate<? super R> rp){ return lp.test(value); }

		@Override
		public Either<L, R> whenLDo(Consumer<L> doF){
			doF.accept(value);
			return this;
		}

		@Override
		public Either<L, R> whenRDo(Consumer<R> doF){ return this; }

		@Override
		public L assumeL(){ return value; }

		@Override
		public L assumeL(Supplier<? extends RuntimeException> toThrowWhenAssumptionInvalid){ return value; }

		@Override
		public R assumeR(){ throw InvalidAlternativeException.left(); }

		@Override
		public R assumeR(Supplier<? extends RuntimeException> toThrowWhenAssumptionInvalid){
			throw toThrowWhenAssumptionInvalid.get();
		}

		@Override
		public boolean equals(Object o){
			if(this == o) return true;
			if(o == null || getClass() != o.getClass()) return false;
			Left<?, ?> left = (Left<?, ?>) o;
			return Objects.equals(value, left.value);
		}

		@Override
		public int hashCode(){ return Objects.hash(Alternative.Left, value); }

		@Override
		public String toString(){ return "Left{" + value + '}'; }
	}

	class Right<L, R> implements Either<L, R> {
		private final R value;

		private Right(R value){ this.value = value; }

		@Override
		public Either<R, L> swap(){ return left(value); }

		@Override
		public <LPrime> Either<LPrime, R> mapL(Function<? super L, LPrime> f){ return right(value); }

		@Override
		public <RPrime> Either<L, RPrime> mapR(Function<? super R, RPrime> f){ return right(f.apply(value)); }

		@Override
		public <LPrime> Either<LPrime, R> flatMapL(Function<? super L, Either<LPrime, R>> f){ return right(value); }

		@Override
		public <RPrime> Either<L, RPrime> flatMapR(Function<? super R, Either<L, RPrime>> f){ return f.apply(value); }

		@Override
		public L collapseIntoL(Function<? super R, ? extends L> f){ return f.apply(value); }

		@Override
		public R collapseIntoR(Function<? super L, ? extends R> f){ return value; }

		@Override
		public Maybe<L> isolateL(){ return none(); }

		@Override
		public Maybe<R> isolateR(){ return just(value); }

		@Override
		public boolean satisfiesL(Predicate<? super L> p){ return false; }

		@Override
		public boolean satisfiesR(Predicate<? super R> p){ return p.test(value); }

		@Override
		public boolean satisfies(Predicate<? super L> lp, Predicate<? super R> rp){ return rp.test(value); }

		@Override
		public Either<L, R> whenLDo(Consumer<L> doF){ return this; }

		@Override
		public Either<L, R> whenRDo(Consumer<R> doF){
			doF.accept(value);
			return this;
		}

		@Override
		public L assumeL(){ throw InvalidAlternativeException.right(); }

		@Override
		public L assumeL(Supplier<? extends RuntimeException> toThrowWhenAssumptionInvalid){
			throw toThrowWhenAssumptionInvalid.get();
		}

		@Override
		public R assumeR(){ return value; }

		@Override
		public R assumeR(Supplier<? extends RuntimeException> toThrowWhenAssumptionInvalid){ return value; }

		@Override
		public boolean equals(Object o){
			if(this == o) return true;
			if(o == null || getClass() != o.getClass()) return false;
			Right<?, ?> right = (Right<?, ?>) o;
			return Objects.equals(value, right.value);
		}

		@Override
		public int hashCode(){ return Objects.hash(Alternative.Right, value); }

		@Override
		public String toString(){ return "Right{" + value + '}'; }
	}

	class InvalidAlternativeException extends RuntimeException {
		private InvalidAlternativeException(Alternative attemptedAlternative){
			super("Attempted to get value of alternative " + attemptedAlternative.name() + " when it was not present" +
			      ".");
		}

		private static InvalidAlternativeException left(){ return new InvalidAlternativeException(Alternative.Left); }

		private static InvalidAlternativeException right(){
			return new InvalidAlternativeException(Alternative.Right);
		}
	}
}

File: src/main/java/io/klbz/curie/Either.kt
package io.klbz.curie

import io.klbz.curie.Maybe.Companion.just
import io.klbz.curie.Maybe.Companion.none
import java.util.function.Consumer
import java.util.function.Function
import java.util.function.Predicate
import java.util.function.Supplier

sealed class Either<L, R> {

    enum class Alternative {
        Left, Right
    }

    abstract fun swap(): Either<R, L>

    abstract fun <LPrime> mapL(f: Function<in L, LPrime>): Either<LPrime, R>

    abstract fun <RPrime> mapR(f: Function<in R, RPrime>): Either<L, RPrime>

    abstract fun <LPrime> flatMapL(f: Function<in L, Either<LPrime, R>>): Either<LPrime, R>

    abstract fun <RPrime> flatMapR(f: Function<in R, Either<L, RPrime>>): Either<L, RPrime>

    abstract fun collapseIntoL(f: Function<in R, out L>): L

    abstract fun collapseIntoR(f: Function<in L, out R>): R

    abstract fun isolateL(): Maybe<L>

    abstract fun isolateR(): Maybe<R>

    abstract fun satisfiesL(p: Predicate<in L>): Boolean

    abstract fun satisfiesR(p: Predicate<in R>): Boolean

    abstract fun satisfies(lp: Predicate<in L>, rp: Predicate<in R>): Boolean

    abstract fun whenLDo(doF: Consumer<L>): Either<L, R>

    abstract fun whenRDo(doF: Consumer<R>): Either<L, R>

    abstract fun assumeL(): L

    abstract fun assumeL(toThrowWhenAssumptionInvalid: Supplier<out RuntimeException>): L

    abstract fun assumeR(): R

    abstract fun assumeR(toThrowWhenAssumptionInvalid: Supplier<out RuntimeException>): R

    //region Kotlin-specific overloads

    @JvmSynthetic
    abstract fun <LPrime> mapL(f: (L) -> LPrime): Either<LPrime, R>

    @JvmSynthetic
    abstract fun <RPrime> mapR(f: (R) -> RPrime): Either<L, RPrime>

    @JvmSynthetic
    abstract fun <LPrime> flatMapL(f: (L) -> Either<LPrime, R>): Either<LPrime, R>

    @JvmSynthetic
    abstract fun <RPrime> flatMapR(f: (R) -> Either<L, RPrime>): Either<L, RPrime>

    @JvmSynthetic
    abstract fun collapseIntoL(f: (R) -> L): L

    @JvmSynthetic
    abstract fun collapseIntoR(f: (L) -> R): R

    @JvmSynthetic
    abstract fun satisfiesL(p: (L) -> Boolean): Boolean

    @JvmSynthetic
    abstract fun satisfiesR(p: (R) -> Boolean): Boolean

    // TODO: Find a more elegant way to phrase this in Kotlin, likely using an intermediate function call
    @JvmSynthetic
    abstract fun satisfies(lp: (L) -> Boolean, rp: (R) -> Boolean): Boolean

    @JvmSynthetic
    abstract fun whenLDo(doF: (L) -> kotlin.Unit): Either<L, R>

    @JvmSynthetic
    abstract fun whenRDo(doF: (R) -> kotlin.Unit): Either<L, R>

    @JvmSynthetic
    abstract fun assumeL(toThrowWhenAssumptionInvalid: () -> RuntimeException): L

    @JvmSynthetic
    abstract fun assumeR(toThrowWhenAssumptionInvalid: () -> RuntimeException): R

    //endregion

    companion object {
        @JvmStatic
        fun <L, R> left(value: L): Either<L, R> = Left(value)

        @JvmStatic
        fun <L, R> right(value: R): Either<L, R> = Right(value)
    }

    internal class Left<L, R> internal constructor(private val value: L) : Either<L, R>() {

        override fun swap() = right<R, L>(value)

        override fun <LPrime> mapL(f: Function<in L, LPrime>) = left<LPrime, R>(f.apply(value))

        override fun <RPrime> mapR(f: Function<in R, RPrime>) = left<L, RPrime>(value)

        override fun <LPrime> flatMapL(f: Function<in L, Either<LPrime, R>>) = f.apply(value)

        override fun <RPrime> flatMapR(f: Function<in R, Either<L, RPrime>>) = left<L, RPrime>(value)

        override fun collapseIntoL(f: Function<in R, out L>) = value

        override fun collapseIntoR(f: Function<in L, out R>) = f.apply(value)

        override fun isolateL() = just(value)

        override fun isolateR() = none<R>()

        override fun satisfiesL(p: Predicate<in L>) = p.test(value)

        override fun satisfiesR(p: Predicate<in R>) = false

        override fun satisfies(lp: Predicate<in L>, rp: Predicate<in R>) = lp.test(value)

        override fun whenLDo(doF: Consumer<L>) = also { doF.accept(value) }

        override fun whenRDo(doF: Consumer<R>) = this

        override fun assumeL() = value

        override fun assumeL(toThrowWhenAssumptionInvalid: Supplier<out RuntimeException>) = value

        override fun assumeR() = throw InvalidAlternativeException.left()

        override fun assumeR(toThrowWhenAssumptionInvalid: Supplier<out RuntimeException>): R {
            throw toThrowWhenAssumptionInvalid.get()
        }

        //region Kotlin-specific overloads

        @JvmSynthetic
        override fun <LPrime> mapL(f: (L) -> LPrime) = left<LPrime, R>(f(value))

        @JvmSynthetic
        override fun <RPrime> mapR(f: (R) -> RPrime) = left<L, RPrime>(value)

        @JvmSynthetic
        override fun <LPrime> flatMapL(f: (L) -> Either<LPrime, R>) = f(value)

        @JvmSynthetic
        override fun <RPrime> flatMapR(f: (R) -> Either<L, RPrime>) = left<L, RPrime>(value)

        @JvmSynthetic
        override fun collapseIntoL(f: (R) -> L) = value

        @JvmSynthetic
        override fun collapseIntoR(f: (L) -> R) = f(value)

        @JvmSynthetic
        override fun satisfiesL(p: (L) -> Boolean) = p(value)

        @JvmSynthetic
        override fun satisfiesR(p: (R) -> Boolean) = false

        @JvmSynthetic
        override fun satisfies(lp: (L) -> Boolean, rp: (R) -> Boolean) = lp(value)

        @JvmSynthetic
        override fun whenLDo(doF: (L) -> kotlin.Unit) = also { doF(value) }

        @JvmSynthetic
        override fun whenRDo(doF: (R) -> kotlin.Unit) = this

        @JvmSynthetic
        override fun assumeL(toThrowWhenAssumptionInvalid: () -> RuntimeException) = value

        @JvmSynthetic
        override fun assumeR(toThrowWhenAssumptionInvalid: () -> RuntimeException) =
                throw toThrowWhenAssumptionInvalid()

        //endregion

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as Left<*, *>

            return value == other.value

        }

        override fun hashCode() = value?.hashCode() ?: 0

        override fun toString(): String = "Left{$value}"
    }

    internal class Right<L, R> internal constructor(private val value: R) : Either<L, R>() {

        override fun swap() = left<R, L>(value)

        override fun <LPrime> mapL(f: Function<in L, LPrime>) = right<LPrime, R>(value)

        override fun <RPrime> mapR(f: Function<in R, RPrime>) = right<L, RPrime>(f.apply(value))

        override fun <LPrime> flatMapL(f: Function<in L, Either<LPrime, R>>) = right<LPrime, R>(value)

        override fun <RPrime> flatMapR(f: Function<in R, Either<L, RPrime>>) = f.apply(value)

        override fun collapseIntoL(f: Function<in R, out L>) = f.apply(value)

        override fun collapseIntoR(f: Function<in L, out R>) = value

        override fun isolateL() = none<L>()

        override fun isolateR() = just(value)

        override fun satisfiesL(p: Predicate<in L>) = false

        override fun satisfiesR(p: Predicate<in R>) = p.test(value)

        override fun satisfies(lp: Predicate<in L>, rp: Predicate<in R>) = rp.test(value)

        override fun whenLDo(doF: Consumer<L>) = this

        override fun whenRDo(doF: Consumer<R>) = also { doF.accept(value) }

        override fun assumeL() = throw InvalidAlternativeException.right()

        override fun assumeL(toThrowWhenAssumptionInvalid: Supplier<out RuntimeException>) =
                throw toThrowWhenAssumptionInvalid.get()

        override fun assumeR() = value

        override fun assumeR(toThrowWhenAssumptionInvalid: Supplier<out RuntimeException>) = value

        //region Kotlin-specific overloads

        @JvmSynthetic
        override fun <LPrime> mapL(f: (L) -> LPrime) = right<LPrime, R>(value)

        @JvmSynthetic
        override fun <RPrime> mapR(f: (R) -> RPrime) = right<L, RPrime>(f(value))

        @JvmSynthetic
        override fun <LPrime> flatMapL(f: (L) -> Either<LPrime, R>) = right<LPrime, R>(value)

        @JvmSynthetic
        override fun <RPrime> flatMapR(f: (R) -> Either<L, RPrime>) = f(value)

        @JvmSynthetic
        override fun collapseIntoL(f: (R) -> L) = f(value)

        @JvmSynthetic
        override fun collapseIntoR(f: (L) -> R) = value

        @JvmSynthetic
        override fun satisfiesL(p: (L) -> Boolean) = false

        @JvmSynthetic
        override fun satisfiesR(p: (R) -> Boolean) = p(value)

        @JvmSynthetic
        override fun satisfies(lp: (L) -> Boolean, rp: (R) -> Boolean) = rp(value)

        @JvmSynthetic
        override fun whenLDo(doF: (L) -> kotlin.Unit) = this

        @JvmSynthetic
        override fun whenRDo(doF: (R) -> kotlin.Unit) = also { doF(value) }

        @JvmSynthetic
        override fun assumeL(toThrowWhenAssumptionInvalid: () -> RuntimeException) =
                throw toThrowWhenAssumptionInvalid()

        @JvmSynthetic
        override fun assumeR(toThrowWhenAssumptionInvalid: () -> RuntimeException) = value

        //endregion

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as Right<*, *>

            return value == other.value

        }

        override fun hashCode() = value?.hashCode() ?: 0

        override fun toString() = "Right{$value}"
    }

    class InvalidAlternativeException private constructor(attemptedAlternative: Alternative)
        : RuntimeException(
            "Attempted to get value of alternative ${attemptedAlternative.name} when it was not present.") {

        companion object {
            internal fun left(): InvalidAlternativeException = InvalidAlternativeException(Alternative.Left)

            internal fun right(): InvalidAlternativeException = InvalidAlternativeException(Alternative.Right)
        }
    }
}
